File: Prototype/Prototy01.js
// Classe Pessoas

class Pessoa{
    constructor(nome, idade){
        this.nome = nome;
        this.idade = idade;
    }

    clone(){
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }
}

// pessoa manager - Deus 

class PessoaManeger{
    constructor(){
        this.pessoas = {};
    }

addPessoa(nome,idade,id){
    const pessoa = new Pessoa(nome,idade);
    this.pessoas[id] = pessoa;  

}

getPessoa(id){
    return this.pessoas[id].clone();
}

}

//
const manager = new PessoaManeger();

//adicionar pessoas
manager.addPessoa("João", 30, 1);
manager.addPessoa("Maria", 25, 2);
manager.addPessoa("Gustavo", 25, 3);
manager.addPessoa("Gilberto", 43, 4);
manager.addPessoa("José", 30, 5);


// clonar pessoa de id=1
const pessoa1 = manager.getPessoa(1);
const pessoa2 = manager.getPessoa(1);


// Modidicar dados da pessoas cloonada
pessoa1.idade = 10;
pessoa2.nome = "João da Silva";
pessoa2.idade = 55;

//Imprime as pessoas

console.log(manager.getPessoa(1))
console.log(manager.getPessoa(2))
console.log(manager.getPessoa(3))
console.log(manager.getPessoa(4))
console.log(manager.getPessoa(5))
console.log("-----------------------")
console.log(pessoa1)
console.log(pessoa2)

File: Prototype/Prototy01.py
import copy
#classe pessoa

class Pessoa:
    def __init__(self,nome,idade):
        self.nome = nome
        self.idade = idade
        
    def __str__ (self):
        return f'Nome: {self.nome}, Idade: {self.idade}'
        
    def clone(self):
        return copy.copy(self)
        
    # classe pessoamangar (Deus)
class PessoaManager:
    def __init__(self):
        self.pessoas = {}
            
    def addPessoa(self, nome, idade, id):
        pessoa = Pessoa(nome, idade)
        self.pessoas[id] = pessoa
            
    def getPessoa(self, id):
        return self.pessoas[id].clone() 
        
        # ultilizar estrutura
manager = PessoaManager()
        
        #adicionar pessoas 
manager.addPessoa("João", 30, 1)
manager.addPessoa("Maria", 25, 2)
manager.addPessoa("Pedro", 20, 3)
manager.addPessoa("Carlos", 30, 4)
manager.addPessoa("Emillia", 50,5)
        
        
        #clonar pessoas
pessoa1 = manager.getPessoa(1)
pessoa2 = manager.getPessoa(5)
        
        #modificar informações 
pessoa1.idade = 18
pessoa2.nome = "alvaro josé"
pessoa2.idade = 55


#imprimir resultados 
print(manager.getPessoa(1))
print(manager.getPessoa(2))
print(manager.getPessoa(3))
print(manager.getPessoa(4))
print(manager.getPessoa(5))
print("------------------------")
print(pessoa1)
print(pessoa2)
               
    